Add dark mode

devnikesh opened the pull request:
## **What**  
I've added support for Dark Mode according to theme preference. It show's up according to the user theme preference in their browser settings. For more background, see ticket [**Add Dark Mode #11**](https://github.com/darekkay/static-marks/issues/11) **Repo:** [darekkay/static-marks](https://github.com/darekkay/static-marks)

## **Why**  
These changes help the users to reduce  reduced glare and blue light exposure and extend battery and was in the issue section `label:Enhancement` 

## **How**  
In App.jsx, we’re checking if the user has set a theme preference in their browser settings. Then addressing them according to the preference by providing theme data to App, which then gets styling through App.scss for Dark and Light mode.

## **Testing**  
 While testing the `window.matchMedia()` method was throwing errors `TypeError: window.matchMedia is not a function` , So I had to add a `matchMedia.js` mock file to pass the Jest Test.

**Links Used to solve the test issue**
* [Mocking methods which are not implemented in JSDOM](https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom)
* [Cannot read property ‘matches’ of undefined](https://github.com/facebook/create-react-app/issues/10126)

## **Screenshots**  

**Test Status**: Passed
![Screenshot (1352)](https://user-images.githubusercontent.com/63302392/149670900-7b0b911b-f397-4483-9394-669b0675740d.png)

**Dark Mode**
![Dark Theme](https://user-images.githubusercontent.com/63302392/149671622-81454364-b3e3-463e-b4b1-e57e6e70d2f1.png)

## **Anything Else?**  
As mentioned in the [**Add Dark Mode #11**](https://github.com/darekkay/static-marks/issues/11) issue, there's also a requirement for Add theme switcher, so we will be adding that functionality in coming PR.
.

darekkay commented on src/App.jsx:
```diff
@@ -38,9 +38,12 @@ const Collections = ({ projects, projectKey, title, currentFilter }) => {
 const App = (props) => {
   const { title, projectKeys, projects, initialFilter } = props;
   const [currentFilter, applyFilter] = useState(initialFilter);
+  const defaultWhite = window.matchMedia(
+    "(prefers-color-scheme: white)"
+  ).matches;
```
NICE: This code can be extracted out of `App` to prevent the recalculation on every rerender.

darekkay commented on src/App.jsx:
```diff
@@ -38,9 +38,12 @@ const Collections = ({ projects, projectKey, title, currentFilter }) => {
   const { title, projectKeys, projects, initialFilter } = props;
   const [currentFilter, applyFilter] = useState(initialFilter);
+  const defaultWhite = window.matchMedia(
+    "(prefers-color-scheme: white)"
```
```suggestion
    "(prefers-color-scheme: light)"
```

See [specs](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme)

Right now, your code does not make it possible to use the light theme. However, after the proposal above, the dark mode is not working anymore:

![image](https://user-images.githubusercontent.com/3101914/149745503-8e380e44-ffc9-4d73-afb6-507ba585962e.png)

Additionally, the current code uses dark mode by default if the use has no preference at all, which I personally don't like as a default value. We should probably invert the logic.

darekkay reviewed (changes requested):
Thanks @devnikesh , this looks like a good work in progress. However, there are some open tasks left:

- Light mode is currently broken, even if the user prefers light mode
- Some parts of the applications still use "light" mode in dark mode (main background, collection names, bucket titles, filter input)
- There are some linter issues, see GitHub action results

devnikesh commented:
@darekkay I apologize for that, I'm on it, the light and dark mode logic is solved on my end, I'm trying to fix the unapplied dark theme on (main background, collection names, bucket titles, filter input)

darekkay commented:
@devnikesh No need to apologize! As mentioned in the main issue, you can probably solve most of this by using the [new atomic color classes](https://github.com/darekkay/static-marks-app/blob/master/src/styles/_atomic.scss) that will match the CSS custom properties. In other words, instead of adjusting CSS and using `var(...)` directly, you can often use classes like `.text-default`, `.text-offset`, `.bg-offset` etc. instead.

devnikesh commented:
@darekkay  yeah sure, will implement this in the next PR, Thank you.

devnikesh commented:
@darekkay  yeah sure, will implement this, Thank you.

devnikesh commented:
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned, hope it works. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

![Dark Theme Enabled](https://user-images.githubusercontent.com/63302392/150006061-18bc43f8-0502-428c-bb72-50bea8abd782.png)

devnikesh commented:
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

![Dark Theme Enabled](https://user-images.githubusercontent.com/63302392/150006061-18bc43f8-0502-428c-bb72-50bea8abd782.png)

devnikesh commented:
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons, Footer titles)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

![Dark Theme Enabled](https://user-images.githubusercontent.com/63302392/150006061-18bc43f8-0502-428c-bb72-50bea8abd782.png)

devnikesh commented:
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons, Footer titles)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

![Dark Theme enabled](https://user-images.githubusercontent.com/63302392/150010058-6dfdc505-cc54-4e88-8711-0bdfe4c914a3.png)

devnikesh commented:
Hi @darekkay I've fixed all those error, also implemented CSS classes as you mentioned. Please have a look at it, and let me know if any changes are required.

**Things done:**

- [x] Fixed the logic error in `App.jsx`
- [x] Extracted color-scheme out of `App`
- [x] Moved the Dark theme code to a dedicated `_dark.scss` file
- [x] Fixed: Enabled Dark theme on (Main background, Collection names, Bucket titles, Filter input, Menu icons, Footer titles)
- [x] Fixed Lint Errors `npm run lint --fix`
- [x] Minor Fixes

**Here's a Screenshot of the Dark Theme :**

![Dark Theme enabled](https://user-images.githubusercontent.com/63302392/150010058-6dfdc505-cc54-4e88-8711-0bdfe4c914a3.png)

darekkay commented:
Thanks @devnikesh ! I'll merge your changes into a feature branch `dark-mode`, so I can do some further alignments with some of my recent work on color schemes.
